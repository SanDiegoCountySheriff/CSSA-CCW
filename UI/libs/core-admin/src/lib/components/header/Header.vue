<template>
  <v-app-bar
    app
    color="accent"
    class="flex-grow-0 white--text"
    clipped-right
  >
    <v-btn
      text
      color="white"
      large
      v-if="authStore.getAuthState.isAuthenticated"
      @click="handleEditAdminUser(false)"
    >
      {{ authStore.getAuthState.userName }}
      <v-icon
        right
        dark
      >
        mdi-cog
      </v-icon>
    </v-btn>

    <v-dialog
      v-model="showAdminUserDialog"
      :persistent="persistentDialog"
      max-width="800px"
    >
      <v-card
        :loading="isCreateAdminUserLoading || isUploadAdminUserDocumentLoading"
      >
        <v-card-title class="headline">
          {{ $t('Setup User Information') }}
        </v-card-title>
        <v-card-text>
          <v-form v-model="valid">
            <v-text-field
              v-model="adminUser.badgeNumber"
              label="Badge Number"
              :rules="[v => !!v || 'Badge Number is required']"
            ></v-text-field>
          </v-form>
          <div class="text-h6">Signature</div>
          <canvas
            width="500px"
            height="100px"
            id="signature"
            class="signature"
          ></canvas>
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn
            color="primary"
            text
            @click="handleClearSignature"
          >
            {{ $t('Clear Signature') }}
          </v-btn>
          <v-btn
            color="primary"
            text
            :disabled="!valid"
            @click="handleSaveAdminUser"
          >
            {{ $t('Save') }}
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>

    <v-spacer></v-spacer>

    <div class="mr-4 ml-1">
      <ThemeMode />
    </div>
    <div
      v-if="
        authStore.getAuthState.isAuthenticated && $vuetify.breakpoint.mdAndUp
      "
      class="caption font-weight-bold mr-4 ml-1"
    >
      {{ $t('Session started at') }} {{ formatTime(sessionTime) }}
    </div>
    <v-btn
      v-if="authStore.getAuthState.isAuthenticated"
      aria-label="Sign out of application"
      @click="signOut"
      class="mr-4 ml-1"
      :color="$vuetify.breakpoint.mdAndDown ? 'accent' : 'accent lighten-2'"
      small
    >
      <!--eslint-disable-next-line vue/singleline-html-element-content-newline -->
      <v-icon
        v-if="$vuetify.breakpoint.mdAndDown"
        class="pr-1 white--text"
      >
        mdi-logout-variant
      </v-icon>
      <span
        v-else
        class="white--text"
        >{{ $t('Sign out') }}</span
      >
    </v-btn>
  </v-app-bar>
</template>

<script setup lang="ts">
import SignaturePad from 'signature_pad';
import ThemeMode from '@shared-ui/components/mode/ThemeMode.vue';
import { UploadedDocType } from '@shared-utils/types/defaultTypes';
import auth from '@shared-ui/api/auth/authentication';
import { formatTime } from '@shared-utils/formatters/defaultFormatters';
import { useAdminUserStore } from '@core-admin/stores/adminUserStore';
import { useAuthStore } from '@shared-ui/stores/auth';
import { useBrandStore } from '@shared-ui/stores/brandStore';
import { useDocumentsStore } from '@core-admin/stores/documentsStore';
import { useMutation } from '@tanstack/vue-query';
import {
  computed,
  nextTick,
  onBeforeUnmount,
  onMounted,
  ref,
  watch,
} from 'vue';

const authStore = useAuthStore();
const brandStore = useBrandStore();
const documentsStore = useDocumentsStore();
const adminUserStore = useAdminUserStore();
const sessionTime = computed(() => authStore.getAuthState.sessionStarted);
const adminUser = computed(() => adminUserStore.adminUser);
const valid = ref(false);
const signaturePad = ref<SignaturePad>();
const persistentDialog = ref(true);
const validAdminUser = ref(adminUserStore.validAdminUser);
const showAdminUserDialog = ref(false);
const form = new FormData();

const { isLoading: isCreateAdminUserLoading, mutate: createAdminUser } =
  useMutation(
    ['createAdminUser'],
    async () => await adminUserStore.putCreateAdminUserApi(adminUser.value),
    {
      onSuccess: async () => {
        showAdminUserDialog.value = false;
        await adminUserStore.getAdminUserApi();
      },
    }
  );

const {
  isLoading: isUploadAdminUserDocumentLoading,
  mutate: uploadAdminUserDocument,
} = useMutation(
  ['uploadAdminUserDocument'],
  async () => await documentsStore.postUploadAdminUserFile(form, 'signature'),
  {
    onSuccess: async () => {
      const uploadDoc: UploadedDocType = {
        documentType: 'adminUserSignature',
        name: 'signature.png',
        uploadedBy: authStore.getAuthState.userName,
        uploadedDateTimeUtc: new Date(Date.now()).toISOString(),
      };

      if (adminUser.value.uploadedDocuments) {
        adminUser.value.uploadedDocuments =
          adminUser.value.uploadedDocuments.filter(document => {
            return document.documentType !== 'adminUserSignature';
          });
      } else {
        adminUser.value.uploadedDocuments = [];
      }

      adminUser.value.uploadedDocuments.push(uploadDoc);

      createAdminUser();
    },
  }
);

let silentRefresh;

onMounted(() => {
  if (authStore.getAuthState.isAuthenticated) {
    silentRefresh = setInterval(
      auth.acquireToken,
      brandStore.getBrand.refreshTokenTime * 1000 * 60
    );
  }

  if (!validAdminUser.value) {
    handleEditAdminUser(true);
  }
});

onBeforeUnmount(() => clearInterval(silentRefresh));

async function signOut() {
  await auth.signOut();
}

function handleEditAdminUser(persist: boolean) {
  persistentDialog.value = persist;
  showAdminUserDialog.value = true;
  nextTick(() => {
    const canvas = document.getElementById('signature') as HTMLCanvasElement;

    signaturePad.value = new SignaturePad(canvas, {
      backgroundColor: 'rgba(255, 255, 255, 0)',
    });

    if (adminUserStore.adminUserSignature) {
      const signature = adminUserStore.adminUserSignature;
      const image = new Image();

      image.src = signature;
      image.onload = () => {
        signaturePad.value?.fromDataURL(signature, {
          ratio: 1,
          width: image.width,
          height: image.height,
        });
      };
    }
  });
}

function handleClearSignature() {
  signaturePad.value?.clear();
}

async function handleSaveAdminUser() {
  const canvas = document.getElementById('signature') as HTMLCanvasElement;

  canvas.toBlob(async blob => {
    form.append('fileToUpload', blob as Blob);

    uploadAdminUserDocument();
  });
}

watch(
  () => validAdminUser.value,
  newVal => {
    if (!newVal) {
      handleEditAdminUser(true);
    }
  }
);
</script>

<style lang="scss" scoped>
.signature {
  border: 2px solid black;
  border-radius: 5px;
}
</style>
