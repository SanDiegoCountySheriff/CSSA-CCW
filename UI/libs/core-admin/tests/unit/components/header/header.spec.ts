/* eslint-disable @typescript-eslint/ban-ts-comment */
import Header from '@core-admin/components/header/Header.vue'
import VueRouter from 'vue-router'
import Vuetify from 'vuetify'
import { createTestingPinia } from '@pinia/testing'
import { createLocalVue, mount } from '@vue/test-utils'

const localVue = createLocalVue()
const pinia = createTestingPinia()
const tMock = {
  $t: text => text,
}

localVue.use(VueRouter)
const router = new VueRouter()

describe('Header', () => {
  let vuetify
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()
    //@ts-ignore
    wrapper = mount(Header, {
      localVue,
      vuetify,
      pinia,
      router,
      mocks: tMock,
    })
  })
  afterEach(() => {
    wrapper.destroy()
  })

  it('should match the snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
