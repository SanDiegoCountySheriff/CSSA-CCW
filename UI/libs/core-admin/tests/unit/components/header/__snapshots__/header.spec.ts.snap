// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot 1`] = `
<header class="flex-grow-0 white--text v-sheet theme--light v-toolbar v-app-bar v-app-bar--clipped v-app-bar--fixed primary" style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;">
  <div class="v-toolbar__content" style="height: 64px;"><i aria-hidden="true" aria-label="Online indicator" aria-haspopup="true" aria-expanded="false" class="v-icon notranslate white--text mdi mdi-wifi theme--light" style="font-size: 22px;"></i><span class="v-tooltip ml-4 v-tooltip--bottom"><!----></span>
    <div class="mode ml-4"><button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--small" aria-label="Change dark/light mode" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate white--text mdi mdi-moon-first-quarter theme--light"></i></span></button><span class="v-tooltip v-tooltip--bottom"><!----></span></div>
    <h3 class="white--text ml-4">
      404
    </h3>
    <div class="spacer"></div>
    <!---->
  </div>
</header>
`;
